domains/client/domain-executor/src/xdm_verifier.rs

```diff
@@ -85,32 +92,35 @@ where
 /// This is used by the System domain to validate Extrinsics.
 /// Returns either true if the XDM is valid else false.
 /// Returns Error when required calls to fetch header info fails.
-pub(crate) fn verify_xdm_with_primary_chain_client<PClient, SClient, PBlock, SBlock>(
+pub(crate) fn verify_xdm_with_primary_chain_client<PClient, PBlock, SBlock, SExecutorDispatch>(
     primary_chain_client: &Arc<PClient>,
-    system_domain_client: &Arc<SClient>,
     extrinsic: &SBlock::Extrinsic,
 ) -> Result<bool, Error>
 where
     PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + 'static,
     PClient::Api: ExecutorApi<PBlock, SBlock::Hash>,
-    SClient: HeaderBackend<SBlock> + ProvideRuntimeApi<SBlock> + 'static,
-    SClient::Api: MessengerApi<SBlock, NumberFor<SBlock>>,
     SBlock: BlockT,
     PBlock: BlockT,
     NumberFor<PBlock>: From<NumberFor<SBlock>>,
     PBlock::Hash: From<SBlock::Hash>,
+    SExecutorDispatch: NativeExecutionDispatch + 'static,
 {
-    let system_domain_runtime = system_domain_client.runtime_api();
-    let best_hash = system_domain_client.info().best_hash;
+    let primary_api = primary_chain_client.runtime_api();
+    let primary_best_hash = primary_chain_client.info().best_hash;
+    let system_domain_wasm = primary_api.system_domain_wasm_bundle(primary_best_hash)?;
+
+    let messenger_api =
+        MessengerApiCompatStateless::<SExecutorDispatch>::with_runtime(system_domain_wasm.to_vec());
```

Comment:
How often is this called? Feels potentially expensive to allocate runtime blob (non-trivial amount of data), maybe worth caching it somewhere rather than instantiating every single time? Substrate internally does caching when you use runtime APIs, but here we're on our own.